code/headers/microphone_controller.hpp

```diff
@@ -0,0 +1,82 @@
+#pragma once
+
+#include "hwlib.hpp"
+
+namespace r2d2::microphone {
+
+    /**
+     * Class microphone_controller_c provides functions
+     * for collecting audio data from the KY-038 module.
+     *
+     * Uses a template to determine the buffer size at
+     * compile time.
+     */
+
+    class microphone_controller_c {
+    private:
+        hwlib::adc &mic;
+	static const int buffer_size = 63; // size of internal communication buffer, this size must not be changed!!
+	uint16_t sample_buffer[buffer_size];
+        bool muted = false; // microphone is not muted by default
+
+    public:
+        /**
+         * Instantiate the microphone controller class
+         *
+         * @param mic The adc input that reads the microphone
+         */
+        microphone_controller_c(hwlib::adc &new_mic):
+	    mic(new_mic)
+	{}
+
+        /**
+         * This function reads the analog data from the
+	 * microphone pin
+         *
+         * @return returns the microphone value
+	 * 
+         */
+        uint16_t read_mic();
+
+	/**
+	 * This function returns a pointer to the sample buffer
+	 *
+	 * @return pointer to sample buffer 
+	 */
+	uint16_t *read_buffer();
+	
+	/**
+	 * This function returns the buffer size
+	 */
+	const int get_buffer_size();
```

Comment:
This function could be marked const. In conjuction with making the buffer size `constexpr`, you could do the following:
```cpp
constexpr int get_buffer_size() const;
```